turbo/snapshotsync/freezeblocks/caplin_state_snapshots.go

```diff
@@ -0,0 +1,667 @@
+				}
+				_, name := filepath.Split(sn.FilePath())
+				if _, ok := protectFiles[name]; ok {
+					continue
+				}
+				toClose = append(toClose, sn)
+			}
+			return true
+		})
+		for _, sn := range toClose {
+			sn.close()
+			segments.DirtySegments.Delete(sn)
+		}
+	}
+}
+
+type CaplinStateView struct {
+	s      *CaplinStateSnapshots
+	roTxs  map[string]*segmentsRotx
+	closed bool
+}
+
+func (s *CaplinStateSnapshots) View() *CaplinStateView {
+	if s == nil {
+		return nil
+	}
+	s.visibleSegmentsLock.RLock()
+	defer s.visibleSegmentsLock.RUnlock()
+
+	v := &CaplinStateView{s: s, roTxs: make(map[string]*segmentsRotx)}
+	// BeginRo increments refcount - which is contended
+	s.dirtySegmentsLock.RLock()
+	defer s.dirtySegmentsLock.RUnlock()
+
+	for k, segments := range s.Segments {
+		v.roTxs[k] = segments.BeginRotx()
+	}
+	return v
+}
+
+func (v *CaplinStateView) Close() {
+	if v == nil {
+		return
+	}
+	if v.closed {
+		return
+	}
+	for _, segments := range v.roTxs {
+		segments.Close()
+	}
+	v.s = nil
+	v.closed = true
+}
+
+func (v *CaplinStateView) VisibleSegments(tbl string) []*VisibleSegment {
+	if v.s == nil || v.s.Segments[tbl] == nil {
+		return nil
+	}
+	return v.s.Segments[tbl].VisibleSegments
+}
+
+func (v *CaplinStateView) VisibleSegment(slot uint64, tbl string) (*VisibleSegment, bool) {
+	for _, seg := range v.VisibleSegments(tbl) {
+		if !(slot >= seg.from && slot < seg.to) {
+			continue
+		}
+		return seg, true
+	}
+	return nil, false
+}
+
+func dumpCaplinState(ctx context.Context, snapName string, kvGetter KeyValueGetter, fromSlot uint64, toSlot, blocksPerFile uint64, salt uint32, dirs datadir.Dirs, workers int, lvl log.Lvl, logger log.Logger, compress bool) error {
+	tmpDir, snapDir := dirs.Tmp, dirs.SnapCaplin
+
+	segName := snaptype.BeaconBlocks.FileName(0, fromSlot, toSlot)
+	// a little bit ugly.
+	segName = strings.ReplaceAll(segName, "beaconblocks", snapName)
+	f, _, _ := snaptype.ParseFileName(snapDir, segName)
+
+	compressCfg := seg.DefaultCfg
+	compressCfg.Workers = workers
+	sn, err := seg.NewCompressor(ctx, fmt.Sprintf("Snapshots %s", snapName), f.Path, tmpDir, compressCfg, lvl, logger)
+	if err != nil {
+		return err
+	}
+	defer sn.Close()
+
+	// Generate .seg file, which is just the list of beacon blocks.
+	for i := fromSlot; i < toSlot; i++ {
+		// read root.
+		_, dump, err := kvGetter(i)
+		if err != nil {
+			return err
+		}
+		if i%20_000 == 0 {
+			logger.Log(lvl, fmt.Sprintf("Dumping %s", snapName), "progress", i)
+		}
+		if compress {
+			if err := sn.AddWord(dump); err != nil {
+				return err
+			}
+		} else {
+			if err := sn.AddUncompressedWord(dump); err != nil {
+				return err
+			}
+		}
+	}
+	if sn.Count() != int(blocksPerFile) {
+		return fmt.Errorf("expected %d blocks, got %d", blocksPerFile, sn.Count())
+	}
+	if err := sn.Compress(); err != nil {
+		return err
+	}
+	// Generate .idx file, which is the slot => offset mapping.
+	p := &background.Progress{}
+
+	// Ugly hack to wait for fsync
```

Comment:
`.Compress()` does wait for `fsync` to finish. you encountered bug? on mac?

Reply:
mmh - it was there before. not sure who put it